Workaround for IME and echo events on Linux (fixes #29 #7106 #9381):

sowfelicity opened the pull request:
Request detectable auto-repeat (Require XKB extension) to support
        echo events and IME at the same time. Fixes #29, #7106 and #9381.

sowfelicity commented:
#9366

Geequlim commented:
Tested with code below
```gdscript
extends Control

func _input(event):
	if event is InputEventKey:
		if event.pressed:
			if not event.is_echo():
				print("Press ", event.unicode)
			else:
				print("Repeat ", event.unicode)
		else:
			print("Release ", event.unicode)
```

* Without active the IME I press and hold then release the A key print as
 ```
Press 97
Repeat 97
Repeat 97
Repeat 97
Repeat 97
Release 97
```
 It works fine with key input with echo.

----------

* Active the IME then press key `A` **twice** and `space` **once** to get a word `啊` with fcitx imput method:
```
Release 97
Release 97
Press 21834
Release 32
```

* Active the IME then press and hold key `A` then press and presee-release `space` to get words `啊啊`  with fcitx imput method:
```
Release 97
Press 21834
Press 21834
Release 32
```

* Active the IME then press and hold key `A` then press-release `Enter` to get words `aaa` with fcitx imput method:
```
Release 97
Press 97
Press 97
Press 97
Release 0
```

* Active the IME then press key `A` **twice** then presse-release key `Esc` to cancel input with fcitx imput method:
```
Release 97
Release 97
Release 0
```

* Active the IME then press and hold key `A` then press-release key `Esc` to cancel input with fcitx imput method:
```
Release 97
Release 0
```

Geequlim commented:
How about ignore all the key events when IME is actived?

Geequlim commented:
@volzhs Would you test the PR with Korean input methods？

volzhs commented:
@Geequlim it works as expected for Korean. :+1:

sowfelicity commented:
It seemed that the type of such event is KeyRelease and there is no KeyPress for the same scancode before.   Such events should be discarded.

sowfelicity commented:
@Geequlim Fixed in X11. These invalid release events will be ignored.

sowfelicity committed:
Workaround for IME and echo events on Linux:
    Request detectable auto-repeat (Require XKB extension) to support
    echo events and IME at the same time. Fixes #29, #7106 and #9381.

akien-mga commented on platform/x11/os_x11.cpp:
```diff
@@ -113,7 +116,21 @@ void OS_X11::initialize(const VideoMode &p_desired, int p_video_driver, int p_au
+	}
+	if (modifiers == NULL) {
+		// To support echo events.
+		WARN_PRINT("IME is disabled");
```
I wouldn't display a warning here, it will confuse the many users who don't even know what an IME is :)
Maybe add to the verbose output instead:
```
if (is_stdout_verbose()) {
	print_line("IME is disabled");
}

sowfelicity replied:
I agree, :). The commit are amended.

akien-mga commented:
To all Linux contributors (both IME users and the rest): please try to build this locally and check that key/echo events work as expected (e.g. test some demos: https://github.com/godotengine/godot-demo-projects/tree/master).

sowfelicity commented:
Here is the [64 bit binary built with gcc 5.4.0](https://github.com/sowfelicity/godot-test/raw/master/godot-linux-64-with-ime-echo.zip). I am not sure whether it can run on all linux distributions.

JunShiozawa commented:
Qt and fcitx-mozc programming hint(2 website).

Japanese language
https://blog.pyyoshi.com/2015/03/03/fcitxhuan-jing-noqt5-4deri-ben-yu-ru-li-dekiruyounisuru/

Google translate of this website(Japanese to English)
https://translate.google.co.jp/translate?sl=ja&tl=en&js=y&prev=_t&hl=ja&ie=UTF-8&u=https%3A%2F%2Fblog.pyyoshi.com%2F2015%2F03%2F03%2Ffcitxhuan-jing-noqt5-4deri-ben-yu-ru-li-dekiruyounisuru%2F&edit-text=

Japanese language
http://diginatu.github.io/ubuntu/fcitx/qt/2015/01/09/qt-fcitx-in-ubuntu.html

Google translate of this website(Japanese to English)
https://translate.google.co.jp/translate?sl=ja&tl=en&js=y&prev=_t&hl=ja&ie=UTF-8&u=http%3A%2F%2Fdiginatu.github.io%2Fubuntu%2Ffcitx%2Fqt%2F2015%2F01%2F09%2Fqt-fcitx-in-ubuntu.html&edit-text=

JunShiozawa commented:
https://github.com/godotengine/godot/blob/master/platform/x11/os_x11.cpp

https://github.com/godotengine/godot/blob/master/platform/x11/godot_x11.cpp

X11 XIM and fcitx-mozc IME programming hint(2 website).

Japanese language
http://rio.la.coocan.jp/lab/xlib/012xim.htm

Google translate of this website(Japanese to English)
https://translate.google.co.jp/translate?sl=ja&tl=en&js=y&prev=_t&hl=ja&ie=UTF-8&u=http%3A%2F%2Frio.la.coocan.jp%2Flab%2Fxlib%2F012xim.htm&edit-text=

Japanese language
http://qiita.com/shibukawa/items/add9a4bb1b1f959aa1fb

Google translate of this website(Japanese to English)
https://translate.google.co.jp/translate?sl=ja&tl=en&js=y&prev=_t&hl=ja&ie=UTF-8&u=http%3A%2F%2Fqiita.com%2Fshibukawa%2Fitems%2Fadd9a4bb1b1f959aa1fb&edit-text=
